### ISSUE TITLE
[Bug] WebGL2RenderingContext leaks when any deck.gl component is removed from DOM

### ISSUE THREAD

**Issue by billyc** [visgl/deck.gl#9264]
### Description

I'm experiencing a memory leak of the entire WebGL2RenderingContext every time I remove a Deck.gl component from the DOM. This is happening when I use both vanilla javascript and the deck.gl react integration. It doesn't matter what layers I insert -- any layer at all will exhibit the behavior (although an empty `layers: [ ]` does not trigger the leak). I have tried both Mapbox and Maplibre map backends, happens with both.

With large datasets this leads to tab crashes quite quickly. 

The documentation is quite clear on how to prevent memory leaks and I've tried to follow everything carefully: I've made sure there are no dangling event listeners; I've attempted `.finalize()` calls etc.  But even the most very basic toy example exhibits this behavior. 

I am attaching a single file HTML which is as small and simple as possible, and it exhibits the problem. 

_I am surely doing something terribly dumb, I can't imagine this leak has gone unnoticed. But I'm not sure what I've done wrong. Happy to provide more info._

### Flavors

- [X] Script tag
- [X] React
- [ ] Python/Jupyter notebook
- [X] MapboxOverlay
- [ ] GoogleMapsOverlay
- [ ] CartoLayer
- [ ] ArcGIS

### Expected Behavior

If I create a deck.gl instance and attach it to a container in the DOM, it shows up properly. When I remove it all resources should be properly cleaned up.

### Steps to Reproduce

A single file JSFiddle here: https://jsfiddle.net/c9ka8vx2/latest/

- Vanilla JavaScript in this test, but I've also tried the react integration
- Toy dataset included in the file

Steps to reproduce:
- Run the fiddle
- Click the "Remove Deck.gl" button
- Open Chrome's Memory developer panel and take a snapshot
- Filter on "Detached" to see a detached WebGL2RenderContext

This can easily leak 100s of megabytes when using large datasets. My users are sad :-(

### Environment

- Framework version: deck.gl@8.9.36
- Browser: Chrome latest
- OS: Macos Ventura and higher

**Comment by ibgreen** [comment]
Yes you are hitting one of the big WebGL limitations. WebGL doesn't offer a method to destroy a WebGL context so we have to rely on garbage collection. And there are only limited amount of contexts that can be created at the same time (8-1.

Anyway you can reuse the canvas element instead of destroying it?

**Comment by ibgreen** [comment]
indeed, destroying all GPU resources before letting go of the WebGL context should help.

**Comment by billyc** [comment]
Now I have to figure out how to get it working in the big internal platform -- but at least now I've proven to myself that the problem isn't in deck.gl -- it's alllll me =)

**Comment by ibgreen** [comment]
TBH, I am not sure what deck.gl is doing to free up the context given that there is no WebGL API to do so. Maybe just more rigorously destroying GPU resources?

Regardless, very glad to hear that it is fixed!

**Comment by billyc** [comment]
Yeah... I completely understand what you're saying here. Probably characterizing it as a "bug" in the 8.x releases is incorrect. I guess there is just something subtle going on with how Deck is interacting with Chrome's garbage collector. But there's nothing subtle about the end effect: now that I have a test case, I can flip it on and off like a light switch. Deck 9.x: no detached WebGL2RenderContext! Deck 8.x: every time.

Anyway I've spent enough time on this. Thank you for your quick responses! 🙏
